roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -65,23 +88,115 @@ impl JobDeclaratorDownstream {
             public_key: config.authority_public_key,
             private_key: config.authority_secret_key,
             mempool,
-            declared_mining_job: None,
+            declared_mining_job: (None, Vec::new(), Vec::new()),
             tx_hash_list_hash: None,
         }
     }
 
-    fn get_block_hex(self_mutex: Arc<Mutex<Self>>, message: SubmitSolutionJd) -> Option<String> {
-        //TODO: implement logic for success or error
-        let (last_declare, tx_list, _) = match self_mutex
-            .safe_lock(|x| x.declared_mining_job.take())
-            .unwrap()
-        {
-            Some((last_declare, tx_list, _x)) => (last_declare, tx_list, _x),
-            None => return None,
-        };
+    // This only errors that are returned are PoisonLock, Custom, MempoolError
+    // this function is called in JobDeclaratorDowenstream::start(), if different errors are
+    // returned, change also the error management there
+    async fn retrieve_transactions_via_rpc(
+        self_mutex: Arc<Mutex<JobDeclaratorDownstream>>,
+    ) -> Result<(), JdsError> {
+        let mut transactions_to_be_retrieved: Vec<Txid> = Vec::new();
+        let mut new_transactions: Vec<Transaction> = Vec::new();
+        let transactions_with_state = self_mutex
+            .clone()
+            .safe_lock(|a| a.declared_mining_job.clone())
+            .map_err(|e| JdsError::PoisonLock(e.to_string()))?
+            .1;
+        for tx_with_state in transactions_with_state {
+            match tx_with_state {
+                TransactionState::Present(_) => continue,
+                TransactionState::ToBeRetrievedFromNodeMempool(tx) => {
+                    transactions_to_be_retrieved.push(tx)
+                }
+                TransactionState::Missing => continue,
+            }
+        }
+        let mempool_ = self_mutex
+            .safe_lock(|a| a.mempool.clone())
+            .map_err(|e| JdsError::PoisonLock(e.to_string()))?;
+        let client = mempool_
+            .clone()
+            .safe_lock(|a| a.get_client())
+            .map_err(|e| JdsError::PoisonLock(e.to_string()))?
+            .ok_or(JdsError::MempoolError(
+                mempool::error::JdsMempoolError::NoClient,
+            ))?;
+        for txid in transactions_to_be_retrieved {
+            let transaction = client
+                .get_raw_transaction(&txid.to_string(), None)
+                .await
+                .map_err(|e| JdsError::MempoolError(mempool::error::JdsMempoolError::Rpc(e)))?;
+            let txid = transaction.txid();
+            mempool::JDsMempool::add_tx_data_to_mempool(
+                mempool_.clone(),
+                txid,
+                Some(transaction.clone()),
```

Comment:
no need to clone the tx data here, just save the id before call add_tx_data_to_mempool and pass the tx data witgout cloning it